src/integrationTest/java/org/opensearch/security/api/RolesRestApiIntegrationTest.java

```diff
@@ -0,0 +1,462 @@
+        );
+        // TODO related to issue #4426
+        for (Iterator<JsonNode> it = expectedObjectNode.get("index_permissions").elements(); it.hasNext();) {
+            final var indexPermission = (ObjectNode) it.next();
+            if (indexPermission.has("dls") && indexPermission.get("dls").isNull()) {
+                indexPermission.remove("dls");
+            }
+            if (indexPermission.has("fls") && indexPermission.get("fls").isNull()) {
+                indexPermission.set("fls", DefaultObjectMapper.objectMapper.createArrayNode());
+            }
+            if (indexPermission.has("masked_fields") && indexPermission.get("masked_fields").isNull()) {
+                indexPermission.set("masked_fields", DefaultObjectMapper.objectMapper.createArrayNode());
+            }
+        }
+
+        assertThat(
+            actualObjectNode.toPrettyString(),
+            actualObjectNode.get("index_permissions"),
+            is(expectedObjectNode.get("index_permissions"))
+        );
+        assertThat(
+            actualObjectNode.toPrettyString(),
+            actualObjectNode.get("tenant_permissions"),
+            is(expectedObjectNode.get("tenant_permissions"))
+        );
+    }
+
+    static ToXContentObject roleWithClusterPermissions(final String... clusterPermissions) {
+        return roleWithClusterPermissions(null, null, null, clusterPermissions);
+    }
+
+    static ToXContentObject roleWithClusterPermissions(
+        final Boolean hidden,
+        final Boolean reserved,
+        final Boolean _static,
+        final String... clusterPermissions
+    ) {
+        return role(
+            hidden,
+            reserved,
+            _static,
+            (builder, params) -> configJsonArray(clusterPermissions).toXContent(builder, params),
+            null,
+            null
+        );
+    }
+
+    static ToXContentObject role(
+        final ToXContentObject clusterPermissions,
+        final ToXContentObject indexPermissions,
+        final ToXContentObject tenantPermissions
+    ) {
+        return role(null, null, null, clusterPermissions, indexPermissions, tenantPermissions);
+    }
+
+    static ToXContentObject role(
+        final Boolean hidden,
+        final Boolean reserved,
+        final ToXContentObject clusterPermissions,
+        final ToXContentObject indexPermissions,
+        final ToXContentObject tenantPermissions
+    ) {
+        return role(hidden, reserved, null, clusterPermissions, indexPermissions, tenantPermissions);
+    }
+
+    static ToXContentObject role(
+        final Boolean hidden,
+        final Boolean reserved,
+        final Boolean _static,
+        final ToXContentObject clusterPermissions,
+        final ToXContentObject indexPermissions,
+        final ToXContentObject tenantPermissions
+    ) {
+        return (builder, params) -> {
+            builder.startObject();
+            builder.field("cluster_permissions");
```

Comment:
nit: This is an existing problem in this repo, but a lot of these Strings end up being repeated throughout the codebase. We may want to consider a single `RoleConstants` class that contains common constants related to roles.

Reply:
True, I'm for it. I will prepare a follow PR.